File: bindings/react-native/src/Statement.ts
/**
 * Statement
 *
 * High-level wrapper around NativeStatement providing a clean API.
 * Handles parameter binding, row conversion, and result collection.
 */

import type {
  NativeConnection,
  NativeStatement,
  SQLiteValue,
  BindParams,
  Row,
  RunResult,
} from './types';
import { TursoStatus, TursoType } from './types';

/**
 * Prepared SQL statement
 */
export class Statement {
  private _statement: NativeStatement;
  private _connection: NativeConnection;
  private _finalized = false;
  private _extraIo?: () => Promise<void>;

  constructor(statement: NativeStatement, connection: NativeConnection, extraIo?: () => Promise<void>) {
    this._statement = statement;
    this._connection = connection;
    this._extraIo = extraIo;
  }

  /**
   * Bind parameters to the statement
   *
   * @param params - Parameters to bind (array, object, or single value)
   * @returns this for chaining
   */
  bind(...params: BindParams[]): this {
    if (this._finalized) {
      throw new Error('Statement has been finalized');
    }

    // Flatten parameters if single array passed
    let flatParams: SQLiteValue[];
    if (params.length === 1 && Array.isArray(params[0])) {
      flatParams = params[0];
    } else if (params.length === 1 && typeof params[0] === 'object' && params[0] !== null) {
      // Named parameters
      const namedParams = params[0] as Record<string, SQLiteValue>;
      this.bindNamed(namedParams);
      return this;
    } else {
      flatParams = params as SQLiteValue[];
    }

    // Bind positional parameters
    this.bindPositional(flatParams);
    return this;
  }

  /**
   * Bind positional parameters (1-indexed)
   *
   * @param params - Array of values to bind
   */
  private bindPositional(params: SQLiteValue[]): void {
    for (let i = 0; i < params.length; i++) {
      const position = i + 1; // 1-indexed
      const value = params[i]!;

      this.bindValue(position, value);
    }
  }

  /**
   * Bind named parameters
   *
   * @param params - Object with named parameters
   */
  private bindNamed(params: Record<string, SQLiteValue>): void {
    for (const [name, value] of Object.entries(params)) {
      // Get position for named parameter
      const position = this._statement.namedPosition(name);
      if (position < 0) {
        throw new Error(`Unknown parameter name: ${name}`);
      }

      this.bindValue(position, value);
    }
  }

  /**
   * Bind a single value at a position
   *
   * @param position - 1-indexed position
   * @param value - Value to bind
   */
  private bindValue(position: number, value: SQLiteValue): void {
    if (value === null || value === undefined) {
      this._statement.bindPositionalNull(position);
    } else if (typeof value === 'number') {
      // Check if integer or float
      if (Number.isInteger(value)) {
        this._statement.bindPositionalInt(position, value);
      } else {
        this._statement.bindPositionalDouble(position, value);
      }
    } else if (typeof value === 'string') {
      this._statement.bindPositionalText(position, value);
    } else if (value instanceof ArrayBuffer || ArrayBuffer.isView(value)) {
      const buffer = value as unknown as ArrayBuffer;
      this._statement.bindPositionalBlob(position, buffer);
    } else {
      throw new Error(`Unsupported parameter type: ${typeof value}`);
    }
  }

  /**
   * Execute statement without returning rows (for INSERT, UPDATE, DELETE)
   *
   * @param params - Optional parameters to bind
   * @returns Result with changes and lastInsertRowid
   */
  async run(...params: BindParams[]): Promise<RunResult> {
    if (this._finalized) {
      throw new Error('Statement has been finalized');
    }

    // Bind parameters if provided
    if (params.length > 0) {
      this.bind(...params);
    }

    // Execute statement with IO handling
    const result = await this.executeWithIo();

    // Reset for next execution
    this._statement.reset();

    return {
      changes: result.rowsChanged,
      lastInsertRowid: this._connection ? this._connection.lastInsertRowid() : 0,
    };
  }

  /**
   * Execute statement handling potential IO (for partial sync)
   * Matches Python's _run_execute_with_io pattern
   *
   * @returns Execution result
   */
  private async executeWithIo(): Promise<{ status: number; rowsChanged: number }> {
    while (true) {
      const result = this._statement.execute();

      if (result.status === TursoStatus.IO) {
        // Statement needs IO (e.g., loading missing pages with partial sync)
        this._statement.runIo();

        // Drain sync engine IO queue
        if (this._extraIo) {
          await this._extraIo();
        }

        continue;
      }

      if (result.status !== TursoStatus.DONE) {
        throw new Error(`Statement execution failed with status: ${result.status}`);
      }

      return result;
    }
  }

  /**
   * Step statement once handling potential IO (for partial sync)
   * Matches Python's _step_once_with_io pattern
   *
   * @returns Status code
   */
  private async stepWithIo(): Promise<number> {
    while (true) {
      const status = this._statement.step();

      if (status === TursoStatus.IO) {
        // Statement needs IO (e.g., loading missing pages with partial sync)
        this._statement.runIo();

        // Drain sync engine IO queue
        if (this._extraIo) {
          await this._extraIo();
        }

        continue;
      }

      return status;
    }
  }

  /**
   * Execute statement and return first row
   *
   * @param params - Optional parameters to bind
   * @returns First row or undefined
   */
  async get(...params: BindParams[]): Promise<Row | undefined> {
    if (this._finalized) {
      throw new Error('Statement has been finalized');
    }

    // Bind parameters if provided
    if (params.length > 0) {
      this.bind(...params);
    }

    // Step once with async IO handling
    const status = await this.stepWithIo();

    if (status === TursoStatus.ROW) {
      const row = this.readRow();
      this._statement.reset();
      return row;
    }

    if (status === TursoStatus.DONE) {
      this._statement.reset();
      return undefined;
    }

    throw new Error(`Statement step failed with status: ${status}`);
  }

  /**
   * Execute statement and return all rows
   *
   * @param params - Optional parameters to bind
   * @returns Array of rows
   */
  async all(...params: BindParams[]): Promise<Row[]> {
    if (this._finalized) {
      throw new Error('Statement has been finalized');
    }

    // Bind parameters if provided
    if (params.length > 0) {
      this.bind(...params);
    }

    const rows: Row[] = [];

    // Step through all rows with async IO handling
    while (true) {
      const status = await this.stepWithIo();

      if (status === TursoStatus.ROW) {
        rows.push(this.readRow());
      } else if (status === TursoStatus.DONE) {
        break;
      } else {
        throw new Error(`Statement step failed with status: ${status}`);
      }
    }

    this._statement.reset();
    return rows;
  }

  /**
   * Read current row into an object
   *
   * @returns Row object with column name keys
   */
  private readRow(): Row {
    const row: Row = {};
    const columnCount = this._statement.columnCount();

    for (let i = 0; i < columnCount; i++) {
      const name = this._statement.columnName(i);
      if (!name) {
        throw new Error(`Failed to get column name at index ${i}`);
      }

      const value = this.readColumnValue(i);
      row[name] = value;
    }

    return row;
  }

  /**
   * Read value at column index
   *
   * @param index - Column index
   * @returns Column value
   */
  private readColumnValue(index: number): SQLiteValue {
    const kind = this._statement.rowValueKind(index);

    switch (kind) {
      case TursoType.NULL:
        return null;

      case TursoType.INTEGER:
        return this._statement.rowValueInt(index);

      case TursoType.REAL:
        return this._statement.rowValueDouble(index);

      case TursoType.TEXT:
        // Use rowValueText which directly returns a string from C++ (avoids encoding issues)
        return this._statement.rowValueText(index);

      case TursoType.BLOB:
        return this._statement.rowValueBytesPtr(index) || new ArrayBuffer(0);

      default:
        throw new Error(`Unknown column type: ${kind}`);
    }
  }

  /**
   * Reset statement for re-execution
   *
   * @returns this for chaining
   */
  reset(): this {
    if (this._finalized) {
      throw new Error('Statement has been finalized');
    }

    this._statement.reset();
    return this;
  }

  /**
   * Finalize and release statement resources
   */
  async finalize(): Promise<void> {
    if (this._finalized) {
      return;
    }

    while (true) {
      const status = this._statement.finalize();

      if (status === TursoStatus.IO) {
        // Statement needs IO (e.g., loading missing pages with partial sync)
        this._statement.runIo();

        // Drain sync engine IO queue
        if (this._extraIo) {
          await this._extraIo();
        }

        continue;
      }

      if (status !== TursoStatus.DONE) {
        throw new Error(`Statement finalization failed with status: ${status}`);
      }
      break;
    }
    this._finalized = true;
  }

  /**
   * Check if statement has been finalized
   */
  get finalized(): boolean {
    return this._finalized;
  }
}

File: bindings/ruby/gem/lib/turso/statement.rb
# frozen_string_literal: true

module Turso
  class Statement
    def initialize(native_statement, native_database)
      @native = native_statement
      @database = native_database
    end

    def bind(*args)
      return self if args.empty?

      if args.size == 1 && args.first.is_a?(Hash)
        bind_hash(args.first)
      else
        args.each_with_index do |value, index|
          bind_value(index + 1, value)
        end
      end
      self
    end

    def run(*bind_args)
      bind(*bind_args) unless bind_args.empty?
      step_loop(consume_rows: false)
      { changes: @database.changes, last_insert_rowid: @database.last_insert_rowid }
    end

    def get(*bind_args)
      bind(*bind_args) unless bind_args.empty?
      row = nil
      step_loop(consume_rows: true) { |r| row ||= r; false }
      row
    end

    def all(*bind_args)
      bind(*bind_args) unless bind_args.empty?
      rows = []
      step_loop(consume_rows: true) { |r| rows << r; true }
      rows
    end

    def each(*bind_args)
      return to_enum(:each, *bind_args) unless block_given?
      bind(*bind_args) unless bind_args.empty?
      step_loop(consume_rows: true) { |r| yield r; true }
    end

    def columns
      (1..@native.column_count).map do |i|
        { name: @native.column_name(i - 1), type: @native.column_decltype(i - 1) }
      end
    end

    def reader?
      @native.column_count > 0
    end

    def step
      @native.step
    end

    def reset
      @native.reset
    end

    def close
      @native.finalize
    end

    private

    def bind_hash(hash)
      hash.each do |key, value|
        name = key.to_s
        name = ":#{name}" unless name.start_with?(/^[:@?$]/)
        position = @native.named_position(name)
        bind_value(position, value)
      end
    end

    def bind_value(position, value)
      case value
      when nil
        @native.bind_null(position)
      when Integer
        @native.bind_int(position, value)
      when Float
        @native.bind_double(position, value)
      when String
        if value.encoding == Encoding::ASCII_8BIT
          @native.bind_blob(position, value.b)
        else
          @native.bind_text(position, value)
        end
      when true
        @native.bind_int(position, 1)
      when false
        @native.bind_int(position, 0)
      when Symbol
        @native.bind_text(position, value.to_s)
      when Time
        @native.bind_text(position, value.iso8601)
      else
        @native.bind_text(position, value.to_s)
      end
    end

    def step_loop(consume_rows:)
      loop do
        result = @native.step
        case result
        when 0 # done
          break
        when 1 # row
          if consume_rows
            row = build_row
            continue = yield row
            break unless continue
          end
        end
      end
    end

    def build_row
      Row.new(@native)
    end
  end
end
